### PR TITLE
Implement orthogonal selection

### PR DESCRIPTION
**Pull request by aleixalcacer**

### CODE REVIEW

File: caterva/caterva.c [hunk 1]
```diff
@@ -1460,3 +1488,420 @@ int caterva_delete(caterva_ctx_t *ctx, caterva_array_t *array, const int8_t axis
...
+            for (int i = 0; i < array->ndim; ++i) {
+                nblock += block_index[i] * block_chunk_strides[i];
+            }
+            maskout[nblock] = false;
+
+            // printf("    - Block: (%lld)", nblock);
+            for (int i = 0; i < array->ndim; ++i) {
+                // printf(" %lld ", block_index[i]);
+            }
+            // printf("\n");
```
Comment (FrancescAlted) [inline caterva/caterva.c]: The block above can be removed.

File: caterva/caterva.c [hunk 2]
```diff
@@ -1460,3 +1488,420 @@ int caterva_delete(caterva_ctx_t *ctx, caterva_array_t *array, const int8_t axis
...
+            int64_t *block_selection_size = malloc(array->ndim * sizeof(int64_t));
+            for (int i = 0; i < array->ndim; ++i) {
+                block_selection_size[i] = chunk_selection_1[i] - chunk_selection_0[i];
+            }
+
+            // printf("    - Block: ");
+            for (int i = 0; i < array->blocknitems; ++i) {
+                // printf(" %f ", ((double *) &data[nblock * array->blocknitems * array->itemsize])[i]);
+            }
+            // printf("\n");
```
Comment (FrancescAlted) [inline caterva/caterva.c]: The block above can be removed

File: caterva/caterva.c [hunk 3]
```diff
@@ -1460,3 +1488,420 @@ int caterva_delete(caterva_ctx_t *ctx, caterva_array_t *array, const int8_t axis
...
+// Indexing
+
+typedef struct {
+    int64_t value;
+    int64_t index;
+} caterva_selection_t;
+
+int caterva_compare_selection(const void * a, const void * b) {
+    int res = (int) (((caterva_selection_t *) a)->value - ((caterva_selection_t *) b)->value);
+    if (res == 0) {
```
Comment (FrancescAlted) [inline caterva/caterva.c:1502]: You could add a comment like: `// In case values are equal, sort by index`

File: caterva/caterva.c [hunk 4]
```diff
@@ -1460,3 +1488,420 @@ int caterva_delete(caterva_ctx_t *ctx, caterva_array_t *array, const int8_t axis
...
+                                   bool get) {
+    p_block_selection_0[ndim] = chunk_selection[ndim];
+    p_block_selection_1[ndim] = chunk_selection[ndim];
+    while (p_block_selection_1[ndim] - p_block_selection_0[ndim] < block_selection_size[ndim]) {
+        if (ndim == array->ndim - 1) {
+            // printf("    - Item: ");
+            for (int i = 0; i < array->ndim; ++i) {
+                // printf(" %lld ", p_block_selection_1[i]->value);
+            }
+            // printf("\n");
```
Comment (FrancescAlted) [inline caterva/caterva.c]: This block can be removed

File: caterva/caterva.c [hunk 5]
```diff
@@ -1460,3 +1488,420 @@ int caterva_delete(caterva_ctx_t *ctx, caterva_array_t *array, const int8_t axis
...
+    p_ordered_selection_0[ndim] = ordered_selection[ndim];
+    p_ordered_selection_1[ndim] = ordered_selection[ndim];
+    while(p_ordered_selection_1[ndim] - ordered_selection[ndim] < selection_size[ndim]) {
+        int64_t chunk_index_ndim = (*p_ordered_selection_1[ndim]).value / array->chunkshape[ndim];
+        while (p_ordered_selection_1[ndim] - ordered_selection[ndim] < selection_size[ndim] &&
+               chunk_index_ndim == (*p_ordered_selection_1[ndim]).value / array->chunkshape[ndim]) {
+            p_ordered_selection_1[ndim]++;
+        }
+        if (ndim == array->ndim - 1) {
+            // printf("- Chunk: ");
```
Comment (FrancescAlted) [inline caterva/caterva.c]: Spare comment:  can be removed

File: caterva/caterva.c [hunk 6]
```diff
@@ -1460,3 +1488,420 @@ int caterva_delete(caterva_ctx_t *ctx, caterva_array_t *array, const int8_t axis
...
+            }
+            int64_t nchunk = 0;
+            for (int i = 0; i < array->ndim; ++i) {
+                nchunk += chunk_index[i] * chunk_array_strides[i];
+            }
+            // printf("(%lld) ", nchunk);
+            for (int i = 0; i < array->ndim; ++i) {
+                // printf(" %lld ", chunk_index[i]);
+            }
+            // printf("\n");
```
Comment (FrancescAlted) [inline caterva/caterva.c]: The block above can be removed

File: caterva/caterva.c [hunk 7]
```diff
@@ -1460,3 +1488,420 @@ int caterva_delete(caterva_ctx_t *ctx, caterva_array_t *array, const int8_t axis
...
+                                                                 chunk_selection_size,
+                                                                 p_ordered_selection_0,
+                                                                 p_chunk_selection_0,
+                                                                 p_chunk_selection_1,
+                                                                 maskout));
+
+                // printf("- Maskout: ");
+                for (int i = 0; i < nblocks; ++i) {
+                    // printf(" %d ", maskout[i]);
+                }
```
Comment (FrancescAlted) [inline caterva/caterva.c]: Ditto

File: caterva/caterva.c [hunk 8]
```diff
@@ -1460,3 +1488,420 @@ int caterva_delete(caterva_ctx_t *ctx, caterva_array_t *array, const int8_t axis
...
+                CATERVA_TRACE_ERROR("Error decompressing chunk");
+                CATERVA_ERROR(CATERVA_ERR_BLOSC_FAILED);
+            }
+            // TODO: Iterate over each chunk and copy data
+            // printf("- Copying data...\n");
+            // printf("    - Chunk data: ");
+            for (int i = 0; i < array->extchunknitems; ++i) {
+                // printf(" %f ", ((double *) data)[i]);
+            }
+            // printf("\n");
```
Comment (FrancescAlted) [inline caterva/caterva.c]: Ditto

File: caterva/caterva.c [hunk 9]
```diff
@@ -1460,3 +1488,420 @@ int caterva_delete(caterva_ctx_t *ctx, caterva_array_t *array, const int8_t axis
...
+    for (int i = 0; i < ndim; ++i) {
+        sel_size *= selection_size[i];
+    }
+
+    if (sel_size < buffersize) {
+        CATERVA_ERROR(CATERVA_ERR_INVALID_ARGUMENT);
+    }
+
+    // Sort selections
+    // printf("Sorting selections...\n");
```
Comment (FrancescAlted) [inline caterva/caterva.c]: Spare comment

File: caterva/caterva.c [hunk 10]
```diff
@@ -1460,3 +1488,420 @@ int caterva_delete(caterva_ctx_t *ctx, caterva_array_t *array, const int8_t axis
...
+
+    for (int i = 0; i < ndim; ++i) {
+        // printf("- Dim %d: |", i);
+        for (int j = 0; j < selection_size[i]; ++j) {
+            // printf(" %lld (%lld) |", ordered_selection[i][j].value, ordered_selection[i][j].index);
+        }
+        // printf("\n");
+    }
+
+    // printf("Selecting chunks...\n");
```
Comment (FrancescAlted) [inline caterva/caterva.c]: Ditto

File: caterva/caterva.c [hunk 11]
```diff
@@ -1460,3 +1488,420 @@ int caterva_delete(caterva_ctx_t *ctx, caterva_array_t *array, const int8_t axis
...
+                                        buffer, buffershape, buffersize, true);
+}
+
+int caterva_set_orthogonal_selection(caterva_ctx_t *ctx, caterva_array_t *array,
+                                     int64_t **selection, int64_t *selection_size,
+                                     void *buffer, int64_t *buffershape, int64_t buffersize) {
+
+    return caterva_orthogonal_selection(ctx, array, selection, selection_size,
+                                        buffer, buffershape, buffersize, false);
+}
```
Comment (FrancescAlted) [inline caterva/caterva.c]: Missing newline at the EOF

**Review by FrancescAlted — APPROVED** [review]
I have a few comments (mainly about cleaning code meant for debugging).

**Review by martaiborra — APPROVED** [review]
LGTM!

**Pull request merged**